### PR TITLE
ODPR-651: Read subscription

### PR DESCRIPTION
**Pull request by paddynski-opencast**

### CODE REVIEW

File: app/services/subscription/SubscriptionService.scala [hunk 1]
```diff
@@ -18,7 +18,7 @@ package services.subscription
 
 object SubscriptionService {
 
-  object Requests {
+  object Data {
```
Comment (erasmos-hmrc-new) [inline app/services/subscription/SubscriptionService.scala:21]: It's a good de-duplication; but please rename it: `RequestOrResponse`.

File: app/connectors/subscription/SubscriptionConnector.scala [hunk 2]
```diff
@@ -16,11 +16,23 @@
...
+
+  final override def connectorPath: String = SubscriptionConnector.connectorPath
+}
 
 object SubscriptionConnector {
-  object Requests {
+
+  val connectorPath: String = "/subscriptions"
+
+  object Data {
```
Comment (erasmos-hmrc-new) [inline app/connectors/subscription/SubscriptionConnector.scala:35]: It's a good de-duplication; but please rename it: `RequestOrResponse`.

File: app/connectors/BaseConnector.scala [hunk 3]
```diff
@@ -62,20 +78,34 @@ abstract class BaseConnector[REQUEST, RESPONSE](wsClient: WSClient) {
   private def asResponse(wsResponse: WSResponse)(implicit reads: Reads[RESPONSE]): Try[Either[Errors, Option[RESPONSE]]] =
     wsResponse.status match {
       case NO_CONTENT => Success(Right(None))
-      case _ =>
-        wsResponse.json
-          .validate[RESPONSE]
-          .map(response => Success(Right(Some(response))))
-          .getOrElse(Failure(new ResponseParsingException()))
+      case _          => Try(
```
Comment (erasmos-hmrc-new) [inline app/connectors/BaseConnector.scala:81]: Good addition.

File: it/test/services/subscription/read/SubscriptionReadServiceSpec.scala [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+              NOT_FOUND,
+              Seq(
+                Error("eis-returned-not-found")
+              )
+            )
+          )
+          verifyThatDownstreamApiWasCalled()
+        }
+      }
+      "fails, with an invalid response body for a status of" - {
```
Comment (erasmos-hmrc-new) [inline it/test/services/subscription/read/SubscriptionReadServiceSpec.scala:265]: I'd suggest the following structure, if only if its more consistent with similar specs - and I don't think this is an improvement on that, but you're welcome to persuade me:

```
"fails, where the response body is" in {
        "absent, with a status code of" in {
          // These would be good to have
          "ok" in {
          }
          "not found" in {

          }
        }
        "valid, with a status code of" - {
          // Contents would be all of: "fails, with a status code of"
        }
        "invalid, with a status code of" - {
          // i.e. it's not valid JSON
        }
        "unexpected, with a status code of" - {
          // i.e. it is valid JSON, but not what expected
          "ok" in {

          }
          "not found" in {

          }
        }
      }
```

File: it/test/services/subscription/read/SubscriptionReadServiceSpec.scala [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+          stubFor(
+            get(urlEqualTo(s"$connectorPath/a7405c8d-06ee-46a3-b5a0-5d65176360ed"))
+              .willReturn(
+                aResponse()
+                  .withHeader("Content-Type", "application/json")
+                  .withStatus(SERVICE_UNAVAILABLE)
+                  .withBody("""
+                              |[
+                              |  {
+                              |    "code": "eis-returned-service-unavailable"
```
Comment (erasmos-hmrc-new) [inline it/test/services/subscription/read/SubscriptionReadServiceSpec.scala:124]: The returned code is actually `eis-returned-internal-server-error`. It's in the ticket, and I just confirmed it [in the backend](https://github.com/hmrc/dprs/blob/2fa5d5f93b465e4277abb12462821f872d402397/it/test/uk/gov/hmrc/dprs/subscription/ReadSubscriptionSpec.scala#L147-L147).

File: it/test/services/subscription/read/SubscriptionReadServiceSpec.scala [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+              )
+          )
+          val request = SubscriptionReadService.Requests.Request(
+            id = "a7405c8d-06ee-46a3-b5a0-5d65176360ed",
+          )
+          val response = await(service.call(request))
+          response shouldBe Left(
+            BaseService.Responses.Errors(
+              INTERNAL_SERVER_ERROR,
+              Seq()
```
Comment (erasmos-hmrc-new) [inline it/test/services/subscription/read/SubscriptionReadServiceSpec.scala:109]: You can just rely on the default for the second arg.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala [hunk 7]
```diff
@@ -17,6 +17,7 @@
 package services.subscription.update
 
 import com.github.tomakehurst.wiremock.client.WireMock._
+import connectors.subscription.SubscriptionConnector
 import connectors.subscription.update.SubscriptionUpdateConnector
```
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:21]: Fixed,

File: it/test/services/subscription/create/SubscriptionCreationServiceSpec.scala [hunk 8]
```diff
@@ -18,6 +18,7 @@ package services.subscription.create
 
 import com.github.tomakehurst.wiremock.client.WireMock._
 import connectors.BaseConnector
+import connectors.subscription.SubscriptionConnector
 import connectors.subscription.create.SubscriptionCreationConnector
```
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply it/test/services/subscription/create/SubscriptionCreationServiceSpec.scala:22]: fixed.

File: app/converters/subscription/SubscriptionConverter.scala [hunk 9]
```diff
@@ -16,9 +16,9 @@
 
 package converters.subscription
 
-import connectors.subscription.SubscriptionConnector.{Requests => ConnectorDataRequests}
```
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply app/converters/subscription/SubscriptionConverter.scala:19]: Ok, Done.

File: app/converters/subscription/SubscriptionConverter.scala [hunk 10]
```diff
@@ -16,9 +16,9 @@
...
+import connectors.subscription.SubscriptionConnector.{Data => ConnectorDataRequests}
 import converters.BaseConverter
-import services.subscription.SubscriptionService.{Requests => ServiceDataRequests}
+import services.subscription.SubscriptionService.{Data => ServiceDataRequests}
```
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply app/converters/subscription/SubscriptionConverter.scala:21]: Ok, Done.

File: app/converters/subscription/update/SubscriptionUpdateConverter.scala [hunk 11]
```diff
@@ -20,18 +20,18 @@ import connectors.BaseConnector
 import connectors.subscription.update.SubscriptionUpdateConnector.{Requests => ConnectorUpdateRequests}
```
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply app/converters/subscription/update/SubscriptionUpdateConverter.scala:20]: Ok, Done.

File: app/connectors/subscription/SubscriptionConnector.scala (hunk 2 above)
Comment (paddynski-opencast) [reply app/connectors/subscription/SubscriptionConnector.scala:35]: Done.

File: it/test/services/subscription/read/SubscriptionReadServiceSpec.scala (hunk 5 above)
Comment (paddynski-opencast) [reply it/test/services/subscription/read/SubscriptionReadServiceSpec.scala]: Reorganised tests.

File: it/test/services/subscription/read/SubscriptionReadServiceSpec.scala (hunk 6 above)
Comment (paddynski-opencast) [reply it/test/services/subscription/read/SubscriptionReadServiceSpec.scala]: It is more ckear what nwe expect this way.

File: it/test/services/subscription/create/SubscriptionCreationServiceSpec.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/create/SubscriptionCreationServiceSpec.scala]: Confirmed.

File: app/converters/subscription/SubscriptionConverter.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/converters/subscription/SubscriptionConverter.scala:19]: Confirmed.

File: app/converters/subscription/update/SubscriptionUpdateConverter.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/converters/subscription/update/SubscriptionUpdateConverter.scala]: Confirmed.

File: app/connectors/subscription/SubscriptionConnector.scala (hunk 2 above)
Comment (erasmos-hmrc-new) [reply app/connectors/subscription/SubscriptionConnector.scala]: Confirmed.

File: it/test/services/subscription/read/SubscriptionReadServiceSpec.scala (hunk 4 above)
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/read/SubscriptionReadServiceSpec.scala]: Confirmed.

File: it/test/services/subscription/read/SubscriptionReadServiceSpec.scala (hunk 6 above)
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/read/SubscriptionReadServiceSpec.scala]: Unchanged.

File: it/test/services/subscription/read/SubscriptionReadServiceSpec.scala (hunk 5 above)
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/read/SubscriptionReadServiceSpec.scala]: Unchanged.

File: app/converters/subscription/read/SubscriptionReadConverter.scala [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package converters.subscription.read
+
+import connectors.BaseConnector
+import connectors.subscription.SubscriptionConnector.{Data => ConnectorDataRequests}
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/converters/subscription/read/SubscriptionReadConverter.scala]: Unchanged.

File: app/converters/subscription/read/SubscriptionReadConverter.scala [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+import connectors.subscription.read
+import connectors.subscription.read.SubscriptionReadConnector
+import converters.subscription.SubscriptionConverter
+import services.BaseService
+import services.subscription.SubscriptionService.{Data => ServiceDataRequests}
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/converters/subscription/read/SubscriptionReadConverter.scala]: Unchanged.

File: app/converters/subscription/read/SubscriptionReadConverter.scala [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package converters.subscription.read
+
+import connectors.BaseConnector
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/converters/subscription/read/SubscriptionReadConverter.scala]: Confirmed.

File: app/converters/subscription/read/SubscriptionReadConverter.scala [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+import connectors.subscription.SubscriptionConnector.{Data => ConnectorDataRequests}
+import connectors.subscription.read
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/converters/subscription/read/SubscriptionReadConverter.scala]: Confirmed.

File: app/connectors/BaseConnector.scala [hunk 16]
```diff
@@ -31,9 +31,25 @@ abstract class BaseConnector[REQUEST, RESPONSE](wsClient: WSClient) {
...
+      .transform {
+        case Success(wsResponse) =>
+          wsResponse.status match {
+            case OK              => asResponse(wsResponse)
+            case otherStatusCode => asErrors(otherStatusCode, wsResponse)
+          }
+        case Failure(exception) => Failure(exception)
+      }
+
   /** We would have liked to use HttpClientV2, but when it encounters a 400 or 500 status code, the response body is inaccessible.
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/connectors/BaseConnector.scala:50]: Confirmed.

File: app/connectors/subscription/SubscriptionConnector.scala [hunk 17]
```diff
@@ -70,5 +102,12 @@ object SubscriptionConnector {
           (JsPath \ "mobile").writeNullable[String] and
           (JsPath \ "emailAddress").write[String])(unlift(Organisation.unapply))
     }
+
+    implicit lazy val reads: Reads[Organisation] =
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/connectors/subscription/SubscriptionConnector.scala:106]: Unchanged.

File: app/connectors/subscription/update/SubscriptionUpdateConnector.scala [hunk 18]
```diff
@@ -18,34 +18,27 @@ package connectors.subscription.update
 
 import com.google.inject.Inject
 import config.FrontendAppConfig
-import connectors.subscription.SubscriptionConnector.Requests.Contact
+import connectors.subscription.SubscriptionConnector
+import connectors.subscription.SubscriptionConnector.Data.Contact
 import connectors.subscription.update.SubscriptionUpdateConnector.Requests.Request
 import connectors.{BaseBackendConnector, BaseConnector}
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/connectors/subscription/update/SubscriptionUpdateConnector.scala]: Confirmed.

File: app/connectors/subscription/create/SubscriptionCreationConnector.scala [hunk 19]
```diff
@@ -18,7 +18,8 @@ package connectors.subscription.create
 
 import com.google.inject.{Inject, Singleton}
 import config.FrontendAppConfig
-import connectors.subscription.SubscriptionConnector.Requests.Contact
+import connectors.subscription.SubscriptionConnector
+import connectors.subscription.SubscriptionConnector.Data.Contact
 import connectors.subscription.create.SubscriptionCreationConnector.Requests.Request
 import connectors.subscription.create.SubscriptionCreationConnector.Responses.Response
 import connectors.{BaseBackendConnector, BaseConnector}
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/connectors/subscription/create/SubscriptionCreationConnector.scala]: Confirmed.

File: app/connectors/subscription/read/SubscriptionReadConnector.scala [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+
+package connectors.subscription.read
+
+import com.google.inject.Inject
+import config.FrontendAppConfig
+import connectors.subscription.SubscriptionConnector
+import connectors.subscription.SubscriptionConnector.Data.Contact
+import connectors.subscription.read.SubscriptionReadConnector.Responses.Response
+import connectors.subscription.read.SubscriptionReadConnector.Requests.Request
+import connectors.{BaseBackendConnector, BaseConnector}
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/connectors/subscription/read/SubscriptionReadConnector.scala]: Confirmed.

File: app/services/subscription/SubscriptionService.scala (hunk 1 above)
Comment (erasmos-hmrc-new) [reply app/services/subscription/SubscriptionService.scala]: Confirmed.

File: it/test/services/subscription/read/SubscriptionReadServiceSpec.scala (hunk 5 above)
Comment (paddynski-opencast) [reply it/test/services/subscription/read/SubscriptionReadServiceSpec.scala]: Fixed.

File: it/test/services/subscription/read/SubscriptionReadServiceSpec.scala (hunk 6 above)
Comment (paddynski-opencast) [reply it/test/services/subscription/read/SubscriptionReadServiceSpec.scala]: Removed empty `Seq()`.

File: app/connectors/BaseConnector.scala (hunk 16 above)
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply app/connectors/BaseConnector.scala:50]: Done.

File: app/connectors/subscription/SubscriptionConnector.scala (hunk 17 above)
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply app/connectors/subscription/SubscriptionConnector.scala:106]: Done.

File: app/converters/subscription/read/SubscriptionReadConverter.scala [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+import connectors.subscription.read.SubscriptionReadConnector
+import converters.subscription.SubscriptionConverter
+import services.BaseService
```
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply app/converters/subscription/read/SubscriptionReadConverter.scala]: ???

File: app/converters/subscription/read/SubscriptionReadConverter.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply app/converters/subscription/read/SubscriptionReadConverter.scala]: Done.

File: app/converters/subscription/read/SubscriptionReadConverter.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply app/converters/subscription/read/SubscriptionReadConverter.scala]: Done.

[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/converters/subscription/read/SubscriptionReadConverter.scala]: Confirmed.

File: app/converters/subscription/read/SubscriptionReadConverter.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/converters/subscription/read/SubscriptionReadConverter.scala]: Confirmed.

File: app/converters/subscription/read/SubscriptionReadConverter.scala (hunk 21 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/converters/subscription/read/SubscriptionReadConverter.scala]: Confirmed.

File: app/connectors/BaseConnector.scala (hunk 16 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/connectors/BaseConnector.scala:52]: Confirmed.

File: app/connectors/subscription/SubscriptionConnector.scala (hunk 17 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/connectors/subscription/SubscriptionConnector.scala]: Confirmed.

File: it/test/services/subscription/read/SubscriptionReadServiceSpec.scala (hunk 5 above)
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/read/SubscriptionReadServiceSpec.scala]: Confirmed.

File: it/test/services/subscription/read/SubscriptionReadServiceSpec.scala (hunk 6 above)
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/read/SubscriptionReadServiceSpec.scala]: I disagree, but resolving as it's a minor point.

**Review by erasmos-hmrc-new — APPROVED** [review]

**Pull request merged**